src/shared/wizards/wizardForm.ts

```diff
@@ -3,87 +3,159 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-import { Prompter } from '../ui/prompter'
 import * as _ from 'lodash'
-import { StateWithCache, WizardState } from './wizard'
-import { ExpandWithObject } from '../utilities/tsUtils'
+import * as toposort from 'toposort'
+import { Prompter } from '../ui/prompter'
+import { StateWithCache } from './wizard'
+import { ExpandWithObject, ReduceTuple } from '../utilities/tsUtils'
+
+export type PrompterProvider<TState, TProp, TDep> = (
+    state: StateWithCache<BoundState<TState, TDep>, TProp>
+) => Prompter<TProp>
+export type BoundState<TState, Bindings> = FindPartialTuple<TState, ExtractPaths<Bindings>>
+
+type DefaultFunction<TState, TProp, TDep> = (state: BoundState<TState, TDep>) => TProp | undefined
+
+/**
+ * Changes all properties enumerated by `K` to be required while leaving everything else
```

Comment:
I'm going off a LOT of faith that this works, I'm assuming there's no way to test this? Seems like a lot of black magic, and unless this is bulletproof I think we'd have a hell of a time debugging it at a later date. Is this absolutely necessary?